pxf/pxf-service/src/main/java/org/apache/hawq/pxf/service/UGICache.java

```diff
@@ -129,90 +135,93 @@ public void release(Entry timedProxyUGI, boolean forceClean) {
     }
 
     /**
-     * Iterate through all the entries in the queue for the given segment
+     * Iterate through all the entries in the queue for the given session's segment
      * and close expired {@link UserGroupInformation}, otherwise it resets
      * the timer for every non-expired entry.
      *
-     * @param segmentId
+     * @param session
      */
-    private void cleanup(Integer segmentId) {
+    private void cleanup(SessionId session) {
 
         Entry ugi;
-        DelayQueue<Entry> delayQueue = getExpirationQueue(segmentId);
+        DelayQueue<Entry> delayQueue = getExpirationQueue(session.getSegmentId());
         while ((ugi = delayQueue.poll()) != null) {
             // Place it back in the queue if still in use and was not closed
-            if (!closeUGI(ugi)) {
+            if (!closeUGI(session, ugi)) {
                 ugi.resetTime();
                 delayQueue.offer(ugi);
             }
             LOG.debug("Delay Queue Size for segment " +
-                    segmentId + " = " + delayQueue.size());
+                    session.getSegmentId() + " = " + delayQueue.size());
         }
     }
 
-    // There is no need to synchronize this method because it should be called
-    // from within a synchronized block
-    private boolean closeUGI(Entry expiredProxyUGI) {
-
-        SessionId session = expiredProxyUGI.getSession();
-        String fsMsg = "FileSystem for proxy user = " + expiredProxyUGI.getUGI().getUserName();
+    /**
+     * This method must be called from a synchronized block for the delayQueue for the given
+     * session.getSegmentId(). When the reference count is 0, the Entry is removed from the
+     * cache where it will then be processed to be destroyed by the UGIProvider.
+     *
+     * @param session
+     * @param toDelete
+     * @return true if the UGI entry was cleaned, false when the UGI entry was still in use
+     * and cleaning up was skipped
+     */
+    private boolean closeUGI(SessionId session, Entry toDelete) {
+        // There is no need to synchronize this method because it should be called
+        // from within a synchronized block
+        String fsMsg = "FileSystem for proxy user = " + toDelete.getUGI().getUserName();
         try {
             // The UGI object is still being used by another thread
-            if (expiredProxyUGI.countReferences() != 0) {
+            if (toDelete.isInUse()) {
                 LOG.info(session.toString() + " Skipping close of " + fsMsg);
-                // Reset time so that it doesn't expire until release
-                // updates the time again
-//                expiredProxyUGI.resetTime();
                 return false;
             }
 
             // Expired UGI object can be cleaned since it is not used
             // Determine if it can be removed from cache also
             Entry cachedUGI = cache.get(session);
-            if (expiredProxyUGI == cachedUGI) {
+            if (toDelete == cachedUGI) {
```

Comment:
Not sure if this logic is still true. The cache could be holding a reference to a newer UGI. That is why we were comparing the reference. This is because the release is also getting it from cache.